httpbp/errors.go

```diff
@@ -0,0 +1,221 @@
+package httpbp
+
+import (
+	"fmt"
+	"html/template"
+	"net/http"
+)
+
+// HTTPError is an error that and can be returned by an  HTTPHandler to return a
+// customized error response.
+type HTTPError interface {
+	error
+
+	// Response returns the custom Response for the error to be written by
+	// the ContentWriter.
+	Response() Response
+
+	// ContentWriter returns the ContentWriter object to use to write the error
+	// response.
+	ContentWriter() ContentWriter
+
+	// Unwrap implements helper interface for errors.Unwrap.  Should return the
+	// internal error that triggered the HTTPError to be returned to the caller.
+	Unwrap() error
+}
+
+// NewHTTPError returns a new HTTPError object initialized with the given
+// values.
+//
+// NewHTTPError is provided for testing purposes and should not be used directly,
+// you should use one of the JSON/HTML/Raw error functions to return standard
+// HTTPErrors
+func NewHTTPError(code int, body interface{}, cause error, cw ContentWriter) HTTPError {
+	return &httpError{
+		resp: Response{
+			Code: code,
+			Body: body,
+		},
+		cw:    cw,
+		cause: cause,
+	}
+}
+
+type httpError struct {
+	resp  Response
+	cw    ContentWriter
+	cause error
+}
+
+func (e httpError) Response() Response {
+	return e.resp
+}
+func (e httpError) ContentWriter() ContentWriter {
+	return e.cw
+}
+
+func (e httpError) Error() string {
+	return fmt.Sprintf(
+		"httpbp: http error with code %d and cause %v",
+		e.Response().Code,
+		e.Unwrap(),
+	)
+}
+
+func (e httpError) Unwrap() error {
+	return e.cause
+}
+
+// ErrorResponse is the base struct used by all the of standard errors in httpbp.
+//
+// ErrorResponse can be used to construct JSON, HTML, and Raw HTTPError responses.
+//
+// You should not generally need to create ErrorResponses manually as standard
+// ones have been provided for each of the common 4xx and 5xx errors.
+type ErrorResponse struct {
+	// The HTTP status code of the response.
+	Code        int
```

Comment:
nit1: I usually prefer to either add json annotation to all fields, or none at all (and rely on go's default behavior to convert field name into json keys). mixing them causes inconsistency.

nit2: since those are not consecutive lines, the align of type won't happen, so you don't need to add extra spaces :)